
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/proto/dataplane/sai";

enum FineGrainedHashFieldAttr {
  FINE_GRAINED_HASH_FIELD_ATTR_UNSPECIFIED = 0;
  FINE_GRAINED_HASH_FIELD_ATTR_NATIVE_HASH_FIELD = 1;
  FINE_GRAINED_HASH_FIELD_ATTR_IPV4_MASK = 2;
  FINE_GRAINED_HASH_FIELD_ATTR_IPV6_MASK = 3;
  FINE_GRAINED_HASH_FIELD_ATTR_SEQUENCE_ID = 4;
}

enum HashAttr {
  HASH_ATTR_UNSPECIFIED = 0;
  HASH_ATTR_NATIVE_HASH_FIELD_LIST = 1;
  HASH_ATTR_UDF_GROUP_LIST = 2;
  HASH_ATTR_FINE_GRAINED_HASH_FIELD_LIST = 3;
}

message CreateFineGrainedHashFieldRequest {
  uint64 switch = 1;

  NativeHashField native_hash_field = 2;
  bytes ipv4_mask = 3;
  bytes ipv6_mask = 4;
  uint32 sequence_id = 5;
}

message CreateFineGrainedHashFieldResponse {
  uint64 oid = 1;
}

message RemoveFineGrainedHashFieldRequest {
  uint64 oid = 1;
}

message RemoveFineGrainedHashFieldResponse {}

message GetFineGrainedHashFieldAttributeRequest {
  uint64 oid = 1;
  FineGrainedHashFieldAttr attr_type = 2;
}

message GetFineGrainedHashFieldAttributeResponse {
  FineGrainedHashFieldAttribute attr = 1;
}

message CreateHashRequest {
  uint64 switch = 1;

  repeated NativeHashField native_hash_field_list = 2;
  repeated uint64 udf_group_list = 3;
  repeated uint64 fine_grained_hash_field_list = 4;
}

message CreateHashResponse {
  uint64 oid = 1;
}

message RemoveHashRequest {
  uint64 oid = 1;
}

message RemoveHashResponse {}

message SetHashAttributeRequest {
  uint64 oid = 1;
  oneof attr {
    NativeHashFieldList native_hash_field_list = 2;
    Uint64List udf_group_list = 3;
    Uint64List fine_grained_hash_field_list = 4;
  }
}

message SetHashAttributeResponse {}

message GetHashAttributeRequest {
  uint64 oid = 1;
  HashAttr attr_type = 2;
}

message GetHashAttributeResponse {
  HashAttribute attr = 1;
}

service Hash {
  rpc CreateFineGrainedHashField(CreateFineGrainedHashFieldRequest)
      returns (CreateFineGrainedHashFieldResponse) {}
  rpc RemoveFineGrainedHashField(RemoveFineGrainedHashFieldRequest)
      returns (RemoveFineGrainedHashFieldResponse) {}
  rpc GetFineGrainedHashFieldAttribute(GetFineGrainedHashFieldAttributeRequest)
      returns (GetFineGrainedHashFieldAttributeResponse) {}
  rpc CreateHash(CreateHashRequest) returns (CreateHashResponse) {}
  rpc RemoveHash(RemoveHashRequest) returns (RemoveHashResponse) {}
  rpc SetHashAttribute(SetHashAttributeRequest)
      returns (SetHashAttributeResponse) {}
  rpc GetHashAttribute(GetHashAttributeRequest)
      returns (GetHashAttributeResponse) {}
}
